### PR TITLE
Handle exception in process_blocklistsubmission task

### LINKED ISSUE #17490: Make sure blocklisting can't get stuck and a submission can't end up in a locked state
On top of #17489, we should make sure that blocks can't get stuck in a phase we can't recover from. I am not sure what the best approach to this is, maybe verify and handle all exceptions, or wrap individual block stages in a transaction so they are rolled back and can be restarted by an admin manually?

### PR DESCRIPTION
**Pull request by eviljeff**
fixes #17490 by rolling back any db transactions; changing the state back to PENDING; and adding a log entry about it

### CODE REVIEW

**Comment by eviljeff** [comment]
cc @wagnerand

**Review by willdurand — COMMENTED** [review]
A quick question while I am trying this patch locally. Everything looks good.

File: src/olympia/blocklist/tasks.py [hunk 1]
```diff
@@ -48,12 +50,27 @@
...
 def process_blocklistsubmission(multi_block_submit_id, **kw):
     obj = BlocklistSubmission.objects.get(pk=multi_block_submit_id)
-    if obj.action == BlocklistSubmission.ACTION_ADDCHANGE:
-        # create the blocks from the guids in the multi_block
-        obj.save_to_block_objects()
-    elif obj.action == BlocklistSubmission.ACTION_DELETE:
-        # delete the blocks
-        obj.delete_block_objects()
+    try:
+        with transaction.atomic():
```
[the remark this answers is not in the document]
Comment (eviljeff) [reply src/olympia/blocklist/tasks.py:54]: It's possible in general, but in this case it'd roll back the update to `obj.signoff_state` and the `LogEntry` explaining the error too.

**Review by willdurand — APPROVED** [review]
Looks good, thanks.

I tried to manually break one of the model methods as you suggested and I noticed `assert` don't have a message by default. Adding a message for each `assert` would be a small improvement given that we log the message of the errors/exceptions in the activity log:

```diff
diff --git a/src/olympia/blocklist/models.py b/src/olympia/blocklist/models.py
index f2fec8f094..43ab72efad 100644
--- a/src/olympia/blocklist/models.py
+++ b/src/olympia/blocklist/models.py
@@ -477,8 +477,8 @@ class BlocklistSubmission(ModelBase):
         }
 
     def save_to_block_objects(self):
-        assert self.is_submission_ready
-        assert self.action == self.ACTION_ADDCHANGE
+        assert self.is_submission_ready, "submission not ready"
+        assert self.action == self.ACTION_ADDCHANGE, "unexpected action"
 
         submit_legacy_switch = waffle.switch_is_active('blocklist_legacy_submit')
         fields_to_set = [

```

**Pull request merged**